pkg/controller/job/job_controller.go

```diff
@@ -812,20 +826,37 @@ func (jm *Controller) manageJob(activePods []*v1.Pod, succeeded int32, job *batc
 			}
 			diff -= batchSize
 		}
-	}
-
-	select {
-	case err := <-errCh:
-		// all errors have been reported before, we only need to inform the controller that there was an error and it should re-try this job once more next time.
-		if err != nil {
+		if err := errorFromChannel(errCh); err != nil {
 			return active, err
 		}
-	default:
 	}
 
 	return active, nil
 }
 
+// activePodsForRemoval returns Pods that should be removed because there
+// are too many pods running or, if this is an indexed job, there are repeated
+// indexes or some pods don't have indexes.
+// Sorts candidate pods in the order such that not-ready < ready, unscheduled
+// < scheduled, and pending < running. This ensures that we delete pods
+// in the earlier stages whenever possible.
+func activePodsForRemoval(pods []*v1.Pod, job *batch.Job, rmAtLeast int) []*v1.Pod {
+	rm := make([]*v1.Pod, 0, rmAtLeast)
+	left := make([]*v1.Pod, 0, len(pods)-rmAtLeast)
```

Comment:
I don't want that function to depend on `rmAtLeast`.

Reply:
I want to avoid doing unnecessary allocation, which seems to be the case for left for non indexed jobs, how about:

```
var rm, left []*v1.Pod
if job.Spec.CompletionMode == batch.IndexedCompletion {
        rm = make([]*v1.Pod, 0, rmAtLeast)
	left = make([]*v1.Pod, 0, len(pods)-rmAtLeast)
	rm, left = appendDuplicatedIndexPodsForRemoval(rm, left, pods)
} else {
	left = pods
}
```

Reply:
oh true, fixed :)